src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/PreferStreamReadAsyncMemoryOverloadsTests.cs

```diff
@@ -0,0 +1,558 @@
+            fs.Read(buffer, 0, (int)fs.Length);
+        }
+    }
+}
+            ");
+        }
+
+        [Fact]
+        public async Task CS_Analyzer_NoDiagnostic_ReadAsync_ByteMemory()
+        {
+            await AnalyzeCSAsync(@"
+using System;
+using System.IO;
+using System.Threading;
+class C
+{
+    public async void M()
+    {
+        using (FileStream fs = new FileStream(""path.txt"", FileMode.Create))
+        {
+            byte[] buffer = new byte[fs.Length];
+            Memory<byte> memory = new Memory<byte>(buffer);
+            await fs.ReadAsync(memory, new CancellationToken()).ConfigureAwait(false);
+        }
+    }
+}
+            ");
+        }
+
+        [Fact]
+        public async Task CS_Analyzer_NoDiagnostic_ReadAsync_AsMemory()
+        {
+            await AnalyzeCSAsync(@"
+using System;
+using System.IO;
+using System.Threading;
+class C
+{
+    public async void M()
+    {
+        using (FileStream fs = File.Open(""file.txt"", FileMode.Open))
+        {
+            byte[] buffer = new byte[fs.Length];
+            await fs.ReadAsync(buffer.AsMemory(), new CancellationToken());
+        }
+    }
+}
+            ");
+        }
+
+        [Fact]
+        public async Task CS_Analyzer_NoDiagnostic_NoAwait_SaveAsTask()
+        {
+            await AnalyzeCSAsync(@"
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+class C
+{
+    public void M()
+    {
+        using (FileStream fs = File.Open(""file.txt"", FileMode.Open))
+        {
+            byte[] buffer = new byte[fs.Length];
+            Task t = fs.ReadAsync(buffer, 0, (int)fs.Length);
+        }
+    }
+}
+            ");
+        }
+
+        [Fact]
+        public async Task CS_Analyzer_NoDiagnostic_NoAwait_ReturnMethod()
+        {
+            await AnalyzeCSAsync(@"
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+class C
+{
+    public Task M(FileStream fs, byte[] buffer)
+    {
+        return fs.ReadAsync(buffer, 0, (int)fs.Length);
+    }
+}
+            ");
+        }
+
+        [Fact]
+        // VB does not support expression bodies for methods
+        public async Task CS_Analyzer_NoDiagnostic_NoAwait_ExpressionBodyMethod()
+        {
+            await AnalyzeCSAsync(@"
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+class C
+{
+    public Task M(FileStream fs, byte[] buffer) => fs.ReadAsync(buffer, 0, (int)fs.Length);
+}
+            ");
+        }
+
+        [Fact]
+        public async Task CS_Analyzer_NoDiagnostic_ConfigureAwait()
+        {
+            await AnalyzeCSAsync(@"
+using System;
+using System.IO;
+using System.Threading;
+class C
+{
+    public async void M()
+    {
+        using (FileStream fs = File.Open(""file.txt"", FileMode.Open))
+        {
+            byte[] buffer = new byte[fs.Length];
+            await fs.ReadAsync(buffer, 0, (int)fs.Length).ConfigureAwait(false);
```

Comment:
I think we want this one to warn...

Reply:
I agree, this we should warn on. We can special-case ConfigureAwait to be as if it didn't exist, for the purposes of this rule.

Reply:
Done.